### ISSUE TITLE
Batches per epoch / training on half batches

### ISSUE THREAD

**Issue by mbernico** [eriklindernoren/Keras-GAN#9]
Hey there...GANs are pretty new to me.  Thank you for all the examples.  

Two questions:

1) Reading your code it seems like you train on one batch per epoch.  Is that intentional?  What's the motivation for that?

2) It seems that training the discriminator on half batches works better than training on a single batch where you concatenate real and fake.  Do you know why that is?

**Comment by eriklindernoren** [comment]
1. Epoch might we a slightly misleading variable name honestly. For each iteration I randomly sample images with uniform probability, so when the number of iterations rises to a sufficiently large number the model will have seen each sample. Therefore I didn't see a reason to divide the training into epochs and batch iterations. MNIST consists of 60k samples, so iterating through the whole dataset during one epoch would take a lot of time. And the models I have implemented that are modelling MNIST converge relatively quickly, so therefore I randomly sample instead.

2. When BatchNormalization2d is used it's recommended to split the data into real and fake batches  before feeding them to the discriminator. It's been a while since I read up on why but since batch norm keeps a running average of batch means and variances I guess mixing batches with data from two distributions messes with these running averages. There are a lot of tricks listed in this great repo: https://github.com/soumith/ganhacks. Dividing these real and fake batches is one trick listed there.

**Comment by mbernico** [comment]
Thanks so much for the help Erik, super cool.

**Comment by woctezuma** [comment]
First, thank you for sharing this code, @eriklindernoren!

Regarding the misleading use of the word _epoch_, it is indeed confusing for newcomers. I guess people who run the code in your repository come from the Keras documentation, so they have [this vocabulary](https://keras.io/getting-started/faq/#what-does-sample-batch-epoch-mean) in mind:

> Epoch: an arbitrary cutoff, generally defined as "one pass over the entire dataset", used to separate training into distinct phases, which is useful for logging and periodic evaluation.

> Batch: a set of N samples. The samples in a batch are processed independently, in parallel. If training, a batch results in only one update to the model.
